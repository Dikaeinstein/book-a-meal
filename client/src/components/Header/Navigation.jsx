import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { removeUser } from '../../helpers/persistUser';
import history from '../../helpers/history';
import { userSignOut } from '../../actions/userActions';

const ConnectedNavigation = ({ urls, signout }) => {
  const handleSignout = (event) => {
    event.preventDefault();
    signout();
    removeUser();
    history.push('/');
  };

  const urlList = urls.map(url => (
    url.link === 'signout'
      ? (
        <li key={url.id}>
          <NavLink
            to={`/${url.link}`}
            onClick={handleSignout}
          >
            {url.name}
          </NavLink>
        </li>
      )
      : (
        <li key={url.id}>
          <NavLink
            to={`/${url.link}`}
          >{url.name}
          </NavLink>
        </li>
      )));

  return (
    <nav className="col-3-4">
      <ul className="nav-container text-center">
        {urlList}
      </ul>
    </nav>
  );
};

const mapStateToProps = state => ({
  urls: state.urls,
});

const mapDispatchToProps = dispatch => ({
  signout() { dispatch(userSignOut()); },
});

ConnectedNavigation.propTypes = {
  urls: PropTypes.arrayOf(PropTypes.object).isRequired,
  signout: PropTypes.func.isRequired,
};

const Navigation = connect(
  mapStateToProps, mapDispatchToProps,
  null, { pure: false },
)(ConnectedNavigation);

export default Navigation;
