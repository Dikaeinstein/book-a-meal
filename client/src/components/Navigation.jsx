import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

const ConnectedNavigation = ({ urls }) => {
  const urlList = urls.map(url => (
    <li key={url.id}>
      <a href={url.link}>{url.name}</a>
    </li>
  ));

  return (
    <nav className="col-3-4">
      <ul className="nav-container text-center">
        {urlList}
      </ul>
    </nav>
  );
};

const mapStateToProps = state => ({
  urls: state.urls,
});

ConnectedNavigation.propTypes = {
  urls: PropTypes.arrayOf(PropTypes.object).isRequired,
};

const Navigation = connect(mapStateToProps)(ConnectedNavigation);

export default Navigation;
