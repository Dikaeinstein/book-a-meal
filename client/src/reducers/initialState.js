const initialState = {
  urls: [
    {
      id: 1,
      name: 'Sign in',
      link: 'signin',
    },
    {
      id: 2,
      name: 'Sign up',
      link: 'signup',
    },
  ],
  menu: {
    isFetching: false,
    data: {
      meals: [],
    },
    error: null,
  },
};

// customerUrls: [
//   {
//     id: 1,
//     name: 'Menu',
//     link: 'user-menu',
//   },
//   {
//     id: 2,
//     name: 'Orders',
//     link: 'user-orderhistory',
//   },
//   {
//     id: 3,
//     name: 'Dikaeinstein',
//     link: 'user-details',
//   },
//   {
//     id: 4,
//     name: 'Logout',
//     link: 'signout',
//   },
// ],
// adminUrls: [
//   {
//     id: 1,
//     name: 'Dashboard',
//     link: 'caterer-dashboard',
//   },
//   {
//     id: 2,
//     name: 'Meals',
//     link: 'meals',
//   },
//   {
//     id: 3,
//     name: 'Menus',
//     link: 'menus',
//   },
//   {
//     id: 4,
//     name: 'Caterer',
//     link: 'user-details',
//   },
//   {
//     id: 5,
//     name: 'Logout',
//     link: 'signout',
//   },
// ],

export default initialState;
