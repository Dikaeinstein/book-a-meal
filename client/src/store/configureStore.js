import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';

import rootReducer from '../reducers';


const getMiddleware = () => {
  const logger = createLogger();
  let middleware = applyMiddleware(thunk);

  if (process.ENV !== 'production') {
    middleware = applyMiddleware(thunk, logger);
  }
  return middleware;
};

const configureStore = initialState => (
  createStore(
    rootReducer,
    initialState,
    compose(
      getMiddleware(),
      window.devToolsExtension ? window.devToolsExtension() : empty => empty,
    ),
  )
);

export default configureStore;
