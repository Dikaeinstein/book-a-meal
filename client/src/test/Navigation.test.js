import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import Navigation from '../components/Navigation';
import configureStore from '../store/configureStore';
import initialState from '../reducers/initialState';
import setup from './setup';

// Setup tests
setup();

const store = configureStore(initialState);

let wrapper;
const urls = [
  {
    id: 1,
    name: 'Sign in',
    link: 'signin',
  },
  {
    id: 2,
    name: 'Sign up',
    link: 'signup',
  },
];

describe('<MockNavigation />', () => {
  beforeAll(() => {
    /* eslint function-paren-newline: 0 */
    wrapper = mount(
      <Provider store={store}>
        <Navigation urls={urls} />
      </Provider>);
  });
  it('should render a Navigation element', () => {
    expect(wrapper.find(Navigation).length).toEqual(1);
  });
  it('should render a ul element', () => {
    expect(wrapper.find('ul').length).toEqual(1);
  });
  it('should render correct number of li elements', () => {
    expect(wrapper.find('li').length).toEqual(urls.length);
  });
  it('should render a tag elements with correct hrefs', () => {
    expect(wrapper.find('a').first().props().href)
      .toEqual(urls[0].link);
    expect(wrapper.find('a').last().props().href)
      .toEqual(urls[1].link);
  });
  it('should match snapshot', () => {
    const tree = mount(
      <Provider store={store}>
        <Navigation urls={urls} />
      </Provider>);
    expect(toJson(tree)).toMatchSnapshot();
  });
});
