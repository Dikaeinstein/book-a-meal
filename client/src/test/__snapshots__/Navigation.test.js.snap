// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MockNavigation /> should match snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(ConnectedNavigation)
    urls={
      Array [
        Object {
          "id": 1,
          "link": "signin",
          "name": "Sign in",
        },
        Object {
          "id": 2,
          "link": "signup",
          "name": "Sign up",
        },
      ]
    }
  >
    <ConnectedNavigation
      dispatch={[Function]}
      urls={
        Array [
          Object {
            "id": 1,
            "link": "signin",
            "name": "Sign in",
          },
          Object {
            "id": 2,
            "link": "signup",
            "name": "Sign up",
          },
        ]
      }
    >
      <nav
        className="col-3-4"
      >
        <ul
          className="nav-container text-center"
        >
          <li
            key="1"
          >
            <a
              href="signin"
            >
              Sign in
            </a>
          </li>
          <li
            key="2"
          >
            <a
              href="signup"
            >
              Sign up
            </a>
          </li>
        </ul>
      </nav>
    </ConnectedNavigation>
  </Connect(ConnectedNavigation)>
</Provider>
`;
